### PR TITLE
CU-2fdg5c0 - Withdraw collateral

### PR DESCRIPTION
**Pull request by 0xangelo**
## Issue
[CU-2fdg5c0](https://app.clickup.com/t/2fdg5c0)

## Description
This implements the functionality described in the ClickUp issue above and its subtasks.

Also includes a bunch of refactorings and feedback from #1144

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2fdg5c0 Feature 7 (Withdraw Collateral Functionality)](https://app.clickup.com/t/2fdg5c0)

**Review by kevin-composable — APPROVED** [review]

**Review by Cardosaum — APPROVED** [review]
LGTM!

Great work in this PR :rocket:

File: frame/clearing-house/src/lib.rs [hunk 1]
```diff
@@ -561,6 +571,55 @@ pub mod pallet {
...
+		/// - [`CollateralWithdrawn`](Event::<T>::CollateralWithdrawn)
+		///
+		/// ## State Changes
+		/// - Settles funding and outstanding profits for all open [`Position`]s
+		/// - Updates the available profits in for the [`Markets`] that the user has open positions
+		///   in
+		/// - Updates the [`Collateral`] of the user
+		///
+		/// The pallet's collateral and insurance accounts are also updated, depending on the amount
+		/// of collateral withdrawn and the bad debt of the system.
```
[the remark this answers is not in the document]
Comment (kevin-composable) [reply frame/clearing-house/src/lib.rs:607]: why not `withdrewn`

[the remark this answers is not in the document]
Comment (kevin-composable) [reply frame/clearing-house/src/lib.rs:607]: jk I think `withdrawn `is ok, but I always get confused with these

File: frame/clearing-house/src/lib.rs [hunk 2]
```diff
@@ -762,7 +821,8 @@ pub mod pallet {
 		/// [`Markets`] is updated, specifically the [`Market`] attributes:
 		/// - [`cum_funding_rate`](Market::<T>::cum_funding_rate)
 		/// - [`funding_rate_ts`](Market::<T>::funding_rate_ts)
-		/// - [`fee_pool`](Market::<T>::fee_pool), if there's Long-Short imbalance
+		///
+		/// The market's Fee Pool account is also updated, if there's Long-Short imbalance
```
[the remark this answers is not in the document]
Comment (0xangelo) [reply frame/clearing-house/src/lib.rs:825]: Goddam periods 😅

File: frame/clearing-house/src/lib.rs [hunk 3]
```diff
@@ -1923,6 +2091,27 @@ pub mod pallet {
...
+		) -> (T::Balance, T::Balance) {
+			let collateral_balance = T::Assets::balance(asset_id, collateral_account);
+			let insurance_balance = T::Assets::balance(asset_id, insurance_account);
+			if amount <= collateral_balance {
+				(amount, T::Balance::zero())
+			} else {
+				(
+					collateral_balance,
+					cmp::min(
+						amount - collateral_balance, /* Safe since amount > collateral_balance */
```
[the remark this answers is not in the document]
Comment (0xangelo) [reply frame/clearing-house/src/lib.rs:2108]: hmm, might be easier to just use the checked version

**Comment by 0xangelo** [comment]
Thanks everyone!

**Pull request merged**